### PR TITLE
Optimize pipeline receive buffer allocation

### PR DESCRIPTION
**Pull request by SherlockNoMad**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #193872

Pipeline receive descriptors allocated activation and gradient buffers for every microbatch during stage initialization and retained them for the whole pipeline step. This reserved memory before communication needed it and required explicit gradient cleanup because the same activation leaves were reused across steps.

Keep only tensor metadata in the receive descriptors, allocate storage immediately before posting an irecv, and transfer the buffer out of the descriptor when microbatch compute consumes it. Same-rank stage handoffs follow the same ownership contract. Fresh receive buffers also make the old input-gradient cleanup and unused non-consuming buffer helper unnecessary.

Test Plan:

```
python test/distributed/pipelining/test_schedule.py -k test_stage_recv_buffers_allocated_just_in_time
```

```
python test/distributed/pipelining/test_schedule_multiproc.py -k test_multi_iter
```

Both test commands reported OK; the test harness exited with status 134 during teardown, matching the existing environment behavior.

## Memory Savings
Configuration: DeepSeek-V3 16B eager, PP4/VPP4, EP2, DP-shard2, microbatch size 1, local batch size 16, sequence length 4096, no activation checkpointing, and 10 training steps on 8 H100 GPUs.

| Metric | `viable/strict` | `pp-lazy-recv-buffers` | Difference |
|---|---:|---:|---:|
| Peak allocated | 66.00 GiB | 64.32 GiB | -1.68 GiB (-2.5%) |
| Peak reserved | 80.25 GiB | 78.59 GiB | -1.66 GiB (-2.1%) |
| GPU headroom | 14.83 GiB | 16.49 GiB | +1.66 GiB |

Lazy receive-buffer allocation reduced peak reserved memory on every PP stage by 1.54-1.66 GiB.

Authored with assistance from an AI coding assistant.

### CODE REVIEW

**Comment by sanketpurandare** [comment]
Superseded by a refreshed current-main ghstack that preserves this implementation and authorship, then adds schedule-derived deterministic receive-buffer pooling. I am taking over maintenance of the replacement stack.

**Comment by sanketpurandare** [comment]
Replacement current-main stack: #196463 carries the preserved lazy-receive implementation, #196464 adds generic schedule-derived resource liveness, and #196465 adds deterministic receive-buffer pooling.
